src/hotspot/share/gc/g1/g1CodeRootSet.cpp

```diff
@@ -143,8 +143,25 @@ class G1CodeRootSetHashTable : public CHeapObj<mtGC> {
   }
 
   void clear() {
+    size_t num_entries = Atomic::load(&_num_entries);
+    if (num_entries != 0) {
+      auto do_value =
+        [] (nmethod** value) {
+          return true;
+      };
+
+      size_t num_deleted = 0;
+      auto do_delete =
+        [&] (nmethod** value) {
+          num_deleted++;
+      };
+
+      bool succeeded = _table.try_bulk_delete(Thread::current(), do_value, do_delete);
+      guarantee(succeeded, "unable to clean table");
+      assert(num_entries == num_deleted, "must empty the table");
+      Atomic::store(&_num_entries, (size_t)0);
+    }
     _table.unsafe_reset();
```

Comment:
Thanks for catching this!

 I believe something like a call to `_table.clean([] (nmethod** value) { return true; });` should do roughly the same. (It is fine to name the lambda with something meaningful, like `always_true`).

Reply:
Another option is to provide a reset function in the CHT itself that roughly does what the destructor does right now.

Reply:
That's the first place I looked. But `free_nodes` is not MT-safe, I am not sure it will fit well with other APIs that all seem to be MT-safe.

I also thought about move `G1CodeRootSetHashTable::clear()` implementation to CHT, but likely I would need another version for safepoint? (e.g. use `_table.bulk_delete` instead).

Hope to have a chance to return to JVM development fulltime, so I will have time to think about it. For now, I will let you to worry about the refactoring :-)

Reply:
I would strongly prefer something like [this](https://github.com/openjdk/jdk/commit/521b66294bc22e7c0760c60e7ce2f1ce31347fc3) after "worrying about the refactoring" ;)

Reply:
It's small enough to copy it in here:
```
  void clear() {
    // Remove all entries.
    auto always_true =
      [] (nmethod** value) {
        return true;
    };

    clean(always_true);
  }
```

Reply:
Ah, I was confused by `_table.clean()` and could not find `clean()` in CHT. Updated.